File: compat/suites/java-sdk/src/main/java/io/overcast/compat/scenario/Json.java
package io.overcast.compat.scenario;

import java.math.BigDecimal;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Map;

/**
 * The IR's document notation: canonical JSON, and equality "as JSON".
 *
 * <p>Every rule the IR states about a response is stated over JSON, so both
 * sides of a comparison are documents by the time they get here — a response
 * through {@link Doc}, an expected value through {@link Binder#eval}, which
 * normalises Java literals the same way. Comparing their canonical encodings is
 * then JSON equality with no coercion: {@code "30"} never equals {@code 30}, and
 * {@code true} never equals {@code 1}.
 *
 * <p>The encoder is written here rather than taken from Jackson because the
 * output is compared and printed, not parsed: object keys must be sorted, HTML
 * must not be escaped, an integral number must render as {@code 30} and not
 * {@code 30.0}, and there must be no trailing newline. Those are the choices
 * that make a java-sdk failure message read like a go-sdk one.
 */
public final class Json {

    private Json() {}

    /**
     * JSON {@code null}, as a value a path can resolve <em>to</em>.
     *
     * <p>Absence and null are different answers from the service and the IR
     * keeps them apart: an absent member is left out of the document entirely,
     * while this singleton is what a null the service really sent resolves to.
     * {@code missing} holds only for the first; {@code nonEmpty} fails for both.
     */
    public static final Object NULL = new Object() {
        @Override
        public String toString() {
            return "null";
        }
    };

    /** What a failure message prints where a path did not resolve. */
    public static final String MISSING = "<missing>";

    /**
     * Renders a document in a stable form: object keys sorted, no HTML
     * escaping, no trailing newline. It is both how values are compared and how
     * they are printed in a failure message, so "expected X, actual Y" reads in
     * the same notation the scenario file is written in.
     */
    public static String canonical(Object v) {
        StringBuilder out = new StringBuilder();
        write(out, v);
        return out.toString();
    }

    /** {@link #canonical}, for a failure message. */
    public static String render(Object v) {
        return canonical(v);
    }

    /** Prints a resolved-or-not value for a failure message. */
    public static String renderResolved(Object v, boolean resolved) {
        return resolved ? render(v) : MISSING;
    }

    /**
     * The IR's "equal, as JSON" ({@code compat/model/README.md} § Assertions).
     */
    public static boolean equal(Object a, Object b) {
        return canonical(a).equals(canonical(b));
    }

    /**
     * The IR's emptiness: null, {@code ""}, {@code []} or <code>{}</code>.
     * Numbers and booleans are never empty, which is what stops {@code nonEmpty}
     * failing on a legitimate 0 or false.
     */
    public static boolean isEmpty(Object v) {
        if (v == null || v == NULL) {
            return true;
        }
        if (v instanceof String s) {
            return s.isEmpty();
        }
        if (v instanceof List<?> l) {
            return l.isEmpty();
        }
        if (v instanceof Map<?, ?> m) {
            return m.isEmpty();
        }
        return false;
    }

    /** Renders one string as a JSON string literal. */
    public static String quote(String s) {
        StringBuilder out = new StringBuilder();
        writeString(out, s);
        return out.toString();
    }

    private static void write(StringBuilder out, Object v) {
        if (v == null || v == NULL) {
            out.append("null");
        } else if (v instanceof String s) {
            writeString(out, s);
        } else if (v instanceof Boolean b) {
            out.append(b ? "true" : "false");
        } else if (v instanceof Number n) {
            writeNumber(out, n);
        } else if (v instanceof Map<?, ?> m) {
            writeObject(out, m);
        } else if (v instanceof List<?> l) {
            writeList(out, l);
        } else {
            // Unreachable for a document Doc produced or a value the emitter
            // wrote: both are limited to the types above. Rendering rather than
            // throwing keeps a failure message readable if one ever is not.
            writeString(out, String.valueOf(v));
        }
    }

    private static void writeObject(StringBuilder out, Map<?, ?> m) {
        List<String> keys = new ArrayList<>(m.size());
        for (Object key : m.keySet()) {
            keys.add(String.valueOf(key));
        }
        Collections.sort(keys);
        out.append('{');
        boolean first = true;
        for (String key : keys) {
            if (!first) {
                out.append(',');
            }
            first = false;
            writeString(out, key);
            out.append(':');
            write(out, m.get(key));
        }
        out.append('}');
    }

    private static void writeList(StringBuilder out, List<?> l) {
        out.append('[');
        for (int i = 0; i < l.size(); i++) {
            if (i > 0) {
                out.append(',');
            }
            write(out, l.get(i));
        }
        out.append(']');
    }

    /**
     * Renders a number the way the interpreters' JSON does: an integral value
     * has no fractional part, whatever Java box it arrived in. Doc normalises
     * every number to a {@code Double}, so without this a response's {@code 30}
     * would print — and compare — as {@code 30.0} against the other backends'
     * {@code 30}.
     */
    private static void writeNumber(StringBuilder out, Number n) {
        double d = n.doubleValue();
        if (Double.isNaN(d) || Double.isInfinite(d)) {
            writeString(out, String.valueOf(n));
            return;
        }
        if (d == Math.rint(d) && Math.abs(d) < 1e15) {
            out.append(Long.toString((long) d));
            return;
        }
        out.append(BigDecimal.valueOf(d).stripTrailingZeros().toPlainString());
    }

    private static void writeString(StringBuilder out, String s) {
        out.append('"');
        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            switch (c) {
                case '"' -> out.append("\\\"");
                case '\\' -> out.append("\\\\");
                case '\n' -> out.append("\\n");
                case '\r' -> out.append("\\r");
                case '\t' -> out.append("\\t");
                case '\b' -> out.append("\\b");
                case '\f' -> out.append("\\f");
                default -> {
                    if (c < 0x20) {
                        out.append(String.format("\\u%04x", (int) c));
                    } else {
                        out.append(c);
                    }
                }
            }
        }
        out.append('"');
    }
}

File: compat/suites/rust-sdk/src/scenario/json.rs
//! Paths, canonical rendering and JSON equality.
//!
//! Paths (compat/model/README.md § Paths): `$` is the response, `.Name` selects
//! a structure member or map key, `[n]` selects a list element. Nothing else —
//! no wildcards, filters, quoting or recursive descent.
//!
//! A path is walked over the response body as it came off the wire (see the
//! module documentation in `mod.rs`), so member names are the modeled names
//! every backend writes and nothing has to be translated between the SDK's
//! object model and the IR's.

use std::fmt::Write as _;

use serde_json::Value as Json;

/// What a failure message prints where a path did not resolve.
pub(crate) const MISSING: &str = "<missing>";

/// One step of a path: a member name or a list index.
enum Segment {
    Member(String),
    Index(usize),
}

/// Splits a path into its segments.
///
/// Anything the IR's path grammar does not admit is rejected, so a malformed
/// path fails the step rather than silently resolving to nothing — the two are
/// very different bugs.
fn parse(path: &str) -> Result<Vec<Segment>, String> {
    let mut rest = match path.strip_prefix('$') {
        Some(rest) => rest,
        None => return Err(format!("path {path:?} does not start with $")),
    };
    let mut segments = Vec::new();
    while !rest.is_empty() {
        if let Some(after) = rest.strip_prefix('.') {
            let end = after.find(['.', '[']).unwrap_or(after.len());
            if end == 0 {
                return Err(format!("path {path:?} has an empty member name"));
            }
            segments.push(Segment::Member(after[..end].to_string()));
            rest = &after[end..];
        } else if rest.starts_with('[') {
            let Some(end) = rest.find(']') else {
                return Err(format!("path {path:?} has an unterminated index"));
            };
            let raw = &rest[1..end];
            match raw.parse::<usize>() {
                Ok(n) => segments.push(Segment::Index(n)),
                Err(_) => return Err(format!("path {path:?} has a non-numeric index {raw:?}")),
            }
            rest = &rest[end + 1..];
        } else {
            let unexpected = rest.chars().next().unwrap_or_default();
            return Err(format!(
                "path {path:?} has an unexpected character {unexpected:?}"
            ));
        }
    }
    Ok(segments)
}

/// Walks a path over a document.
///
/// `Ok(None)` means a segment was absent — which is what `missing` tests for,
/// and what makes an absent list count as empty for `listContains` and
/// `absent`. A member the service sent as JSON `null` is present and resolves
/// to `null`; a member it omitted does not resolve, and neither does an index
/// past the end of a list.
pub(crate) fn resolve<'a>(doc: &'a Json, path: &str) -> Result<Option<&'a Json>, String> {
    let mut current = doc;
    for segment in parse(path)? {
        match segment {
            Segment::Index(n) => match current.as_array().and_then(|items| items.get(n)) {
                Some(next) => current = next,
                None => return Ok(None),
            },
            Segment::Member(name) => match current.as_object().and_then(|obj| obj.get(&name)) {
                Some(next) => current = next,
                None => return Ok(None),
            },
        }
    }
    Ok(Some(current))
}

/// Renders a value in a stable form: object keys sorted, no trailing newline.
///
/// It is both how values are printed in a failure message and how they are
/// compared, so "expected X, actual Y" reads in the same notation the scenario
/// file is written in. The keys are sorted here rather than left to
/// `serde_json`, whose ordering depends on a cargo feature another crate in the
/// graph could turn on.
pub(crate) fn render(value: &Json) -> String {
    let mut out = String::new();
    write_canonical(&mut out, value);
    out
}

fn write_canonical(out: &mut String, value: &Json) {
    match value {
        Json::Null => out.push_str("null"),
        Json::Bool(b) => {
            let _ = write!(out, "{b}");
        }
        Json::Number(n) => {
            let _ = write!(out, "{n}");
        }
        Json::String(s) => out.push_str(&Json::String(s.clone()).to_string()),
        Json::Array(items) => {
            out.push('[');
            for (i, item) in items.iter().enumerate() {
                if i > 0 {
                    out.push(',');
                }
                write_canonical(out, item);
            }
            out.push(']');
        }
        Json::Object(entries) => {
            let mut keys: Vec<&String> = entries.keys().collect();
            keys.sort();
            out.push('{');
            for (i, key) in keys.into_iter().enumerate() {
                if i > 0 {
                    out.push(',');
                }
                out.push_str(&Json::String(key.clone()).to_string());
                out.push(':');
                write_canonical(out, &entries[key]);
            }
            out.push('}');
        }
    }
}

/// Renders a resolved-or-not value for a failure message.
pub(crate) fn render_resolved(value: Option<&Json>) -> String {
    match value {
        Some(value) => render(value),
        None => MISSING.to_string(),
    }
}

/// The IR's "equal, as JSON" (compat/model/README.md § Assertions).
///
/// Comparison is in the JSON type system directly, with no coercion: `"30"`
/// never equals `30` and `true` never equals `1`. Numbers are the one place a
/// normalisation is needed — a service that answers `30` and one that answers
/// `30.0` have said the same thing, and `serde_json` keeps the two
/// representations apart — so numbers are compared as `f64`, which is also what
/// the three interpreters compare after parsing.
pub(crate) fn equal(a: &Json, b: &Json) -> bool {
    match (a, b) {
        (Json::Number(x), Json::Number(y)) => match (x.as_f64(), y.as_f64()) {
            (Some(x), Some(y)) => x == y,
            _ => x == y,
        },
        (Json::Array(x), Json::Array(y)) => {
            x.len() == y.len() && x.iter().zip(y).all(|(x, y)| equal(x, y))
        }
        (Json::Object(x), Json::Object(y)) => {
            x.len() == y.len()
                && x.iter()
                    .all(|(key, value)| y.get(key).is_some_and(|other| equal(value, other)))
        }
        _ => a == b,
    }
}

/// The IR's emptiness: `null`, `""`, `[]` or `{}`. Numbers and booleans are
/// never empty, which is what stops `nonEmpty` failing on a legitimate 0 or
/// `false`.
pub(crate) fn is_empty(value: &Json) -> bool {
    match value {
        Json::Null => true,
        Json::String(s) => s.is_empty(),
        Json::Array(items) => items.is_empty(),
        Json::Object(entries) => entries.is_empty(),
        _ => false,
    }
}
